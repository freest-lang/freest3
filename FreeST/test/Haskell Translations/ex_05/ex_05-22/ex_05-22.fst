-- V exercise 22

data IntList = End | List Int IntList

selectApply : (Int -> Int) -> (Int -> Bool) -> IntList -> IntList
selectApply applier iffer list =
    case list of {
        End -> End,
        List x rest ->
            if iffer x
                then List (applier x) (selectApply applier iffer rest)
                else (selectApply applier iffer rest)
    }

list : IntList
list = List (-4) (List (-3) (List (-2) (List (-1) 
       (List 0 (List 1 (List 2 (List 3 (List 4 End))))))))

main : IntList
main = selectApply (\x : Int -> x*3) (\x : Int -> x > 0) list
--result = List 3 (List 6 (List 9 (List 12 End)))
