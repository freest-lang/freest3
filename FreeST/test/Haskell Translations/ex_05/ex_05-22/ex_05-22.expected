List 3 (List 6 (List 9 (List 12 End)))